nodejs/dist/uws.js

```diff
@@ -1,98 +1,132 @@
-        }
+
+const uws = require(`./uws_${process.platform}_${process.versions.modules}`);
+const NativeServer = uws.Server;
+
+class Socket extends EventEmitter {
+    /**
+     * Creates a Socket instance.
+     *
+     * @param {Object} nativeSocket Socket instance
+     * @param {Server} server Server instance
+     */
+    constructor(nativeSocket, server) {
+        super();
+        this.nativeSocket = nativeSocket;
+        this.server = server;
     }
 
-    this.close = function() {
-        if (self.nativeSocket !== null) {
-            self.server.nativeServer.close(self.nativeSocket);
-            self.nativeSocket = null;
-        } else {
-            /* ignore close on closed sockets */
-        }
+    /**
+     * Sends a message.
+     *
+     * @param {String|Buffer} message The message to send
+     * @param {Object} options Send options
+     * @public
+     */
+    send(message, options) {
+        /* ignore sends on closed sockets */
+        if (!this.nativeSocket) return;
+
+        const binary = options && options.binary || false;
+        this.server.nativeServer.send(this.nativeSocket, message, binary);
+    }
+
+    /**
+     * Closes the socket.
+     *
+     * @public
+     */
+    close() {
+        /* ignore close on closed sockets */
+        if (!this.nativeSocket) return;
+
+        this.server.nativeServer.close(this.nativeSocket);
+        this.server = this.nativeSocket = undefined;
     }
 }
 
-module.exports.Server = function Server(options) {
-    EventEmitter.call(this);
+class Server extends EventEmitter {
+    /**
+     * Creates a Server instance.
+     *
+     * @param {Object} options Configuration options
+     */
+    constructor(options) {
+        super();
+        /* we currently do only support port as option */
+        this.nativeServer = new NativeServer(options.port);
+
+        this.nativeServer.onConnection((nativeSocket) => {
+            const socket = new Socket(nativeSocket, this);
+            this.nativeServer.setData(nativeSocket, socket);
+            this.emit('connection', socket);
+        });
+
+        this.nativeServer.onDisconnection((nativeSocket) => {
+            const socket = this.nativeServer.getData(nativeSocket);
+            socket.emit('close');
+            /* make sure to clear any set data */
+            this.nativeServer.setData(nativeSocket);
+        });
 
-    /* we currently do only support port as option */
-    this.nativeServer = new NativeServer(options.port);
-    var self = this;
+        this.nativeServer.onMessage((nativeSocket, message, binary) => {
+            const socket = this.nativeServer.getData(nativeSocket);
+            socket.emit('message', binary ? message : message.toString());
+        });
+    }
 
-    this.close = function () {
-        self.nativeServer.close();
+    /**
+     * Closes the server.
+     *
+     * @public
+     */
+    close() {
+        this.nativeServer.close();
+        this.nativeServer = undefined;
```

Comment:
I meant this line, the close is not closing the sockets this instant, so you cannot set nativeServer to undefined.